peps/pep-0735.rst

```diff
@@ -0,0 +1,476 @@
+(PDM also uses extras for some Dependency Groups, and overlaps the notion
+heavily with extras.)
+
+This PEP is not guaranteed to be a perfectly substitutable solution for the
+same problem space for each tool. However, the ideas are extremely similar, and
+it should be possible for Poetry and PDM to support at least some
+PEP-735-standardized Dependency Group configurations using their own Dependency
+Group nomenclature.
+
+A level of interoperability with Poetry and PDM is a goal of this PEP, but
+certain features and behaviors defined here may not be supported by Poetry and
+PDM. Matching the existing Poetry and PDM *semantics* for Dependency Groups is
+a non-goal.
+
+Dependency Groups are not Hidden Extras
+---------------------------------------
+
+One could be forgiven for thinking that Dependency Groups are just extras which
+go unpublished.
+
+However, there are two major features they have which distinguish them from
+extras:
+
+* they support non-package projects
+
+* installation of a dependency group does not imply installation of a package's
+  dependencies (or the package itself)
+
+Specification
+=============
+
+This PEP defines a new section (table) in ``pyproject.toml`` files named
+``requirements``. The ``requirements`` table contains exactly one key,
+``packages``, which is a table. All other keys in ``requirements`` are reserved
+for future use.
+
+The ``packages`` table contains an arbitrary number of user-defined keys, each of
+which has, as its value, a list of requirements specifiers (defined below).
+These keys must match the following regular expression:
+``[a-z0-9][a-z0-9-]*[a-z0-9]``. Meaning that they must be all lower-case
+alphanumerics, with ``-`` allowed only in the middle, and at least
+two characters long. These requirements are chosen so that the normalization
+rules used for pypi package names do not apply, as the names are already
+normalized.
+
+Requirements specifiers will use a definition which extends PEP 508. Therefore,
+it is first necessary to define a string format, a "PEP 735 Dependency". These
+are defined as strings in one of the three following formats:
+
+* a PEP 508 specification, e.g., ``numpy>1``
+* a comma-delimited list of Dependency Group names, enclosed in square
+  brackets, e.g., ``[test,docs]`` to refer to other Dependency Groups named
+  ``test`` and ``docs``
+* a single dot, ``.``, which refers to the current project as a package
+  (similar to ``pip install .``)
+* a single dot, followed by square brackets enclosing a list of extra names,
+  which refers to the current project as a package including some extras, e.g.,
+  ``.[mysql]`` to refer to the current package with its ``mysql`` extra
+
+Requirement specifiers can now be defined as one of the following:
+
+* A string, which is a valid PEP 735 Dependency Specifier. e.g., ``numpy>1``
+* An object, which has exactly one key, ``spec``, which is a valid PEP 735
+  Dependency Spec. e.g., ``{spec = "numpy>1"}``
```

Comment:
Where did this come from? The phrasing of, "can now be defined" suggests you already discussed all the relevant details, but this is the first mention of anything well beyond requirement specifiers.

And what's the motivation for this object addition? Nowhere in the motivation or rationale did anything get put forth which would suggest this is necessary. I would argue it goes against "the desire for this format to be as simple and learnable as possible" (maybe this was discussed on discuss.python.org and I haven't read the discussion yet?).

Reply:
> The phrasing of, "can now be defined" suggests you already discussed all the relevant details, but this is the first mention of anything well beyond requirement specifiers.

Yeah, bad/misleading phrasing.
I've cut it down to "Requirement specifiers are one of the following:" for now. (will push a new rev shortly)

> And what's the motivation for this object addition? Nowhere in the motivation or rationale did anything get put forth which would suggest this is necessary. I would argue it goes against "the desire for this format to be as simple and learnable as possible" (maybe this was discussed on discuss.python.org and I haven't read the discussion yet?).

This is my own invention. It's being discussed on discourse now because I've just introduced it, but you're not missing any background discussion.

I missed it from the Rationale section -- I need to add it. Even if the object format gets nixed and we/I remove it, the current text is less coherent without including relevant rationale.
My goal was to specify how tools should behave if new fields are added in the future, and to explicitly make it possible for new fields to be added by starting with the string/object duality.
My concern was that if the object format isn't introduced from the outset, it can't be safely introduced later without causing a lot of breakage/churn.
If it's built in from the start, fields can be added over time without having to come back and alter this specification.

The format is inspired by [PEP 633](https://peps.python.org/pep-0633/). Reading the discussion of that PEP, it seemed to me that it was rejected largely because (1) it proposed many fields at once, and (2) it uses the PEP 440 directly for `version`, rather than using full PEP 508 strings.

I might be mistaken in this whole line of thinking. I can be convinced. But I was worried about leaving this out of the spec and potentially regretting it later if people want support for pip's `--hash` option, in particular.

Reply:
Over the past two iterations of this document, I think the changes have evolved enough that this particular comment can be marked resolved.

Even if `only-deps` and `editable` are removed, I would like to keep path dependencies and includes declared as objects. As with everything, I'm open to convincing arguments, but for the purposes of this first draft, I think the PEP does a reasonable job of arguing in favor of objects and against a special syntax.